controllers/resources/lease.go

```diff
@@ -0,0 +1,233 @@
+		return timeout
+	}
+	return DefaultLeaseDuration
+}
+
+func (m *nhcLeaseManager) getTimeoutForRemediation(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Duration {
+	var leaseDuration time.Duration
+	remediationKind := remediationCR.GetKind()
+	for _, esRemediation := range nhc.Spec.EscalatingRemediations {
+		if strings.TrimSuffix(esRemediation.RemediationTemplate.Kind, "Template") == remediationKind {
+			leaseDuration = esRemediation.Timeout.Duration
+			break
+		}
+	}
+	return leaseDuration
+}
+
+func (m *nhcLeaseManager) isLeaseOverdue(l *coordv1.Lease, nhc *remediationv1alpha1.NodeHealthCheck, remediationCR *unstructured.Unstructured) (bool, error) {
+	if l.Spec.AcquireTime == nil {
+		err := fmt.Errorf("lease Spec.AcquireTime is nil")
+		m.log.Error(err, "lease Spec.AcquireTime is nil", "lease name", l.Name)
+		return false, err
+	}
+
+	isLeaseOverdue := time.Now().After(m.calcLeaseExpiration(l, remediationCR, nhc))
+	return isLeaseOverdue, nil
+}
+
+func (m *nhcLeaseManager) calcLeaseExpiration(l *coordv1.Lease, remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Time {
+	leaseDuration := m.getLeaseDurationForRemediation(remediationCR, nhc)
+	prevRemediationsDuration := m.sumPrevRemediationsDuration(remediationCR, nhc)
+	return l.Spec.AcquireTime.Add(time.Duration(maxTimesToExtendLease+1 /*1 is offsetting the lease creation*/)*leaseDuration + prevRemediationsDuration)
+}
+
+func (m *nhcLeaseManager) isRemediationsExist(remediationCrs []unstructured.Unstructured) bool {
+	return len(remediationCrs) > 0
+}
+
+func (m *nhcLeaseManager) isLeaseOwner(l *coordv1.Lease) bool {
+	if l.Spec.HolderIdentity == nil {
+		return false
+	}
+	return *l.Spec.HolderIdentity == holderIdentity
+}
+
+func (m *nhcLeaseManager) sumPrevRemediationsDuration(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Duration {
+	if len(nhc.Spec.EscalatingRemediations) == 0 {
+		return 0
+	}
+
+	var currentEscalatingRemediation *remediationv1alpha1.EscalatingRemediation
+	remediationKind := remediationCR.GetKind()
+	for _, esRemediation := range nhc.Spec.EscalatingRemediations {
+		if strings.TrimSuffix(esRemediation.RemediationTemplate.Kind, "Template") == remediationKind {
+			currentEscalatingRemediation = &esRemediation
+			break
+		}
+	}
+	//remediation isn't found in escalation - should not happen
+	if currentEscalatingRemediation == nil {
+		msg := fmt.Sprintf("existing remediation of kind:%s not found in escalation list: %v", remediationCR.GetKind(), nhc.Spec.EscalatingRemediations)
+		m.log.Error(fmt.Errorf("remediation not found in escalation"), msg)
+		return 0
+	}
+
+	var sumDurations time.Duration
+	for _, esRemediation := range nhc.Spec.EscalatingRemediations {
+		if currentEscalatingRemediation.Order > esRemediation.Order {
+			sumDurations += esRemediation.Timeout.Duration
```

Comment:
Just for the record, this calculation isn't accurate when a remediation failed early by setting the succeeded condition to false. But good enough, expiring later than intended isn't an issue.